account_invoice_pricelist/models/account_move.py

```diff
@@ -0,0 +1,238 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    pricelist_id = fields.Many2one(
+        comodel_name="product.pricelist",
+        string="Pricelist",
+        readonly=True,
+        states={"draft": [("readonly", False)]},
+    )
+
+    @api.constrains("pricelist_id", "currency_id")
+    def _check_currency(self):
+        for sel in self.filtered(lambda a: a.pricelist_id and a.is_invoice()):
+            if sel.pricelist_id.currency_id != sel.currency_id:
+                raise UserError(
+                    _("Pricelist and Invoice need to use the same currency.")
+                )
+
+    @api.onchange("partner_id", "company_id")
+    def _onchange_partner_id_account_invoice_pricelist(self):
+        if self.is_invoice():
+            result = super(AccountMove, self)._onchange_partner_id()
+            if (
+                self.partner_id
+                and self.type in ("out_invoice", "out_refund")
+                and self.partner_id.property_product_pricelist
+            ):
+                self.pricelist_id = self.partner_id.property_product_pricelist
+                self._calculate_currency()
+            return result
+
+    @api.onchange("pricelist_id")
+    def _calculate_currency(self):
```

Comment:
```suggestion
    def _set_pricelist_currency(self):
```
Non-blocking. IMHO, to understand what the code does. WDYT?

Reply:
Done